corehq/util/workbook_json/excel.py

```diff
@@ -188,6 +190,11 @@ def _convert_float(value):
 class WorkbookJSONReader(object):
 
     def __init__(self, file_or_filename):
+        if isinstance(file_or_filename, InMemoryUploadedFile):
+            tmp = NamedTemporaryFile(mode='wb', suffix='.xlsx', delete=False)
```

Comment:
This file should get deleted automatically - can the `load_workbook` bit be put inside a `with NamedTemporaryFile` or `TransientTempfile` block?

I think fine to PR separately if we're trying to get this in to today's deploy

Reply:
thank you @esoergel .
Tried updating this first to see how feasible this was
```
    def __init__(self, file_or_filename):
        try:
            if isinstance(file_or_filename, InMemoryUploadedFile):
                with NamedTemporaryFile(mode='wb+', suffix='.xlsx') as tmp:
                    tmp.write(file_or_filename.read())
                    self.wb = openpyxl.load_workbook(tmp, read_only=True, data_only=True)
                    self._load_workbook()
            else:
                self.wb = openpyxl.load_workbook(file_or_filename, read_only=True, data_only=True)
                self._load_workbook()
        except (BadZipfile, InvalidFileException, KeyError) as e:
            raise InvalidExcelFileException(six.text_type(e))

    def _load_workbook(self):
        self.worksheets_by_title = {}
        self.worksheets = []

        for worksheet in self.wb.worksheets:
            try:
                ws = WorksheetJSONReader(worksheet, title=worksheet.title)
            except IndexError:
                raise JSONReaderError('This Excel file has unrecognised formatting. Please try downloading '
                                      'the lookup table first, and then add data to it.')
            self.worksheets_by_title[worksheet.title] = ws
            self.worksheets.append(ws)
```

Everything here went fine, but it failed later with `ValueError: I/O operation on closed file` when it was iterating rows in `get_unicode_dicts`.
trace here
```
  File "/Users/Manish/Projects/commcare-hq/corehq/apps/app_manager/app_translations/app_translations.py", line 195, in validate_bulk_app_translation_upload
    msgs = UploadedTranslationsValidator(app, workbook).compare()
  File "/Users/Manish/Projects/commcare-hq/corehq/apps/app_manager/app_translations/validator.py", line 81, in compare
    rows = get_unicode_dicts(sheet)
  File "/Users/Manish/Projects/commcare-hq/corehq/apps/app_manager/app_translations/app_translations.py", line 50, in get_unicode_dicts
    for row in iterable:
  File "/Users/Manish/Projects/commcare-hq/corehq/util/workbook_json/excel.py", line 61, in __iter__
    for row in self._rows:
  File "/Users/Manish/Projects/commcare-hq/corehq/util/workbook_json/excel.py", line 179, in iterator
    for row in self.worksheet.iter_rows():
  File "/Users/Manish/.virtualenvs/commcare-hq/lib/python2.7/site-packages/openpyxl/worksheet/read_only.py", line 120, in _cells_by_row
    for _event, element in p:
  File "<string>", line 100, in next
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 632, in read
    data = self.read1(n - len(buf))
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 665, in read1
    data = self._fileobj.read(nbytes)
ValueError: I/O operation on closed file
```
So looks like it needs the file around after its parsed here.

Reply:
That's a shame - it must not read the whole thing into memory.  The proper fix would I think be to move the resource allocation high enough up the call stack that all operations on that tempfile are encapsulated.  I'm okay dropping it though, since it's quite outside the scope of this PR.

Reply:
Sure. Thanks for the follow up @esoergel